ssl/s3_lib.c

```diff
@@ -4148,6 +4151,34 @@ const SSL_CIPHER *ssl3_choose_cipher(SSL *s, STACK_OF(SSL_CIPHER) *clnt,
     if (s->options & SSL_OP_CIPHER_SERVER_PREFERENCE || tls1_suiteb(s)) {
         prio = srvr;
         allow = clnt;
+#ifndef OPENSSL_NO_CHACHA
+        /* If ChaCha is at the top of the client preference list,
+           temporarily prioritize ChaCha ciphers in the servers list. */
+        if (s->options & SSL_OP_PRIORITIZE_CHACHA && sk_SSL_CIPHER_num(clnt) > 0) {
+            c = sk_SSL_CIPHER_value(clnt, 0);
+            if (c->algorithm_enc == SSL_CHACHA20POLY1305) {
+                /* Is client preferred, reprioritize... */
+                new_chacha = sk_SSL_CIPHER_new_null();
+                if (new_chacha != NULL) {
+                    /* Put ChaCha at the top */
+                    for (i = 0; i < sk_SSL_CIPHER_num(srvr); i++) {
+                        c = sk_SSL_CIPHER_value(srvr, i);
+                        if (c->algorithm_enc == SSL_CHACHA20POLY1305) {
+                            sk_SSL_CIPHER_push(new_chacha, c);
+                        }
+                    }
+                    /* Pull in the rest */
```

Comment:
Why does one have to compose the rest of the list?

Reply:
Although both lists may have ChaCha20, they may not be the same cipher suite; so the rest of the cipher list is necessary. This prevents a failure case (no matching cipher) that would have otherwise succeeded.

Reply:
(I have seen other ChaCha20 priority patches that have this failure mode because they don't consider the non-ChaCha ciphers their processing)

Reply:
Git it.

Reply:
:-) I missed one centimeter too! I mean I meant to write "got it"...